### PR TITLE
Storybook Composition - loading a prebuilt storybook into yours

### PR DESCRIPTION
**Pull request by ndelangen**
Storybook is used by teams, and it's actually quite common for multiple teams to work on their own isolated features, possibly in their own storybook. But it's also common these teams will share components between them.

In fact a very common scenario is that you're using some open source components, maybe a component library. They are using storybook, you are using storybook. But if you want to see their components -and the documentation that come with it- you have to go to their website/storybook or write stories for components that aren't yours.

This is wasting time. Storybook should have a mechanism to deal with this.

Presenting: storybook inception!

```js
// main.js
module.exports = {
  stories: [`${__dirname}/stories/*.*`],
  addons: [
    '@storybook/addon-roundtrip/register',
    '@storybook/addon-parameter/register',
    '@storybook/addon-preview-wrapper/register',
  ],
  refs: {
    inception: 'https://storybookjs-next.now.sh/dev-kits/iframe.html',
  },
};
```

The new `refs` lists all external (pre-build) storybooks you want storybook to load.

The storybook is loaded, and the stories defined in both will appear in the navigator.

Here's one of the challenges I faced: Stories might have the same name, and thus overwrite each other.
The default way storybook now deals with this is by prefixing all stories's kind with the id of the ref that loaded it.

There's a way to tweak this behaviour, because I expect the consumers will want some control over where included stories are placed. Here's how you do it:

```js
// manager.js
import { addons } from '@storybook/addons';

addons.setConfig({
  mapper: (refObject, storyObject) => {
    // return a storyObject, perhaps with a kind you customized
    return storyObject;
  },
});
```

If you return `null` then that story will be ignored, and not shown in the navigator.

---

The stories from 1 ref and another are completely isolated. Storybook will actually not merge them together, in fact storybook will load the actual iframes as specified in the ref, and 'juggle' multiple iframes and show the correct iframe for the currently selected story,

Currently this process isn't optimised, which means all refs are loaded from the start.
Once storybook has a way to pull the list of stories statically, we can lazy-load the iframe.

---

todo still

- [ ] expandable refs
- [ ] test auth flow
- [ ] documentation
- [ ] extract command with lazy puppeteer install
- [ ] preset to auto-load refs from dependencies

### CODE REVIEW

File: lib/api/src/index.tsx [hunk 1]
```diff
@@ -178,7 +185,46 @@ class ManagerProvider extends Component<Props, State> {
...
+                ...mapped,
+                id: `${id}_${mapped.id}`,
+                knownAs: k,
+                source: u,
+              };
+            }
+          });
+        }
+      } else {
+        console.log('no refs loaded');
```
Comment (shilman) [inline lib/api/src/index.tsx]: ???

**Comment by shilman** [comment]
Hey @ndelangen, would love to add a more "convention over configuration" API. Let's discuss offline before merging. In the meantime will test out what you currently have!

File: lib/api/src/index.tsx (hunk 1 above)
Comment (ndelangen) [reply lib/api/src/index.tsx]: Should be removed

File: lib/api/src/index.tsx [hunk 2]
```diff
@@ -178,7 +185,44 @@ class ManagerProvider extends Component<Props, State> {
...
+              data.stories[`${id}_${mapped.id}`] = {
+                ...mapped,
+                id: `${id}_${mapped.id}`,
+                knownAs: k,
+                source: u,
+              };
+            }
+          });
+        }
+      }
```
Comment (bostanio) [inline lib/api/src/index.tsx]: I find the code above super dense and it's kinda hard to read and grok what its doing. Can we break it up into a few named methods, add comments  etc. to make it more obvious what the logic is?

Comment (ndelangen) [reply lib/api/src/index.tsx]: I've tried to make it simpler, WDYT?

**Comment by ndelangen** [comment]
@shilman let's talk about this in the near future.

File: lib/addons/src/index.ts [hunk 3]
```diff
@@ -3,36 +3,21 @@ import global from 'global';
 import { ReactElement } from 'react';
```
Comment (ndelangen) [inline lib/addons/src/index.ts]: I should cherry-pick this file change to `next`

File: lib/api/src/init-provider-api.ts [hunk 4]
```diff
@@ -17,7 +20,12 @@ export interface Provider {
   channel?: Channel;
   renderPreview?: IframeRenderer;
   handleAPI(api: API): void;
-  getConfig(): Record<string, any>;
+  getConfig(): {
```
Comment (ndelangen) [inline lib/api/src/init-provider-api.ts]: I could pull some of this out into a separate PR

File: lib/api/src/lib/stories.ts [hunk 5]
```diff
@@ -2,40 +2,52 @@ import deprecate from 'util-deprecate';
...
+  knownAs?: StoryId;
   depth: 0;
   name: string;
   children: StoryId[];
   isComponent: false;
   isRoot: true;
   isLeaf: false;
   // MDX stories are "Group" type
-  parameters?: any;
+  parameters?: {
```
Comment (ndelangen) [inline lib/api/src/lib/stories.ts]: This TS improvement could be extracted as well

File: lib/api/src/modules/layout.ts [hunk 6]
```diff
@@ -56,7 +56,7 @@ type PartialThemeVars = Partial<ThemeVars>;
 type PartialLayout = Partial<Layout>;
 type PartialUI = Partial<UI>;
 
-interface Options extends ThemeVars {
+export interface UIOptions {
```
Comment (ndelangen) [inline lib/api/src/modules/layout.ts]: This isn't Inception specific either, extract to it's own PR

File: lib/addons/src/index.ts (hunk 3 above)
Comment (ndelangen) [reply lib/addons/src/index.ts]: ✔️

File: lib/api/src/init-provider-api.ts (hunk 4 above)
Comment (ndelangen) [reply lib/api/src/init-provider-api.ts]: ✔️

File: lib/api/src/lib/stories.ts (hunk 5 above)
Comment (ndelangen) [reply lib/api/src/lib/stories.ts]: ✔️

File: lib/api/src/modules/layout.ts (hunk 6 above)
Comment (ndelangen) [reply lib/api/src/modules/layout.ts]: ✔️

File: lib/api/src/index.tsx [hunk 7]
```diff
@@ -197,19 +212,76 @@ class ManagerProvider extends Component<ManagerProviderProps, State> {
...
+        case 'local': {
+          api.selectStory(kind, story, rest);
+          break;
+        }
+
+        case 'external': {
+          const refs = api.getRefs();
+          const [refId] = Object.entries(refs).find(([, { url }]) =>
+            `${url}/iframe.html`.match(source)
+          );
```
Comment (ndelangen) [inline lib/api/src/index.tsx]: Seems shared logic

File: lib/api/src/lib/stories.ts [hunk 8]
```diff
@@ -25,9 +28,11 @@ export interface Root {
 
 export interface Group {
   id: StoryId;
+  knownAs?: StoryId;
```
Comment (ndelangen) [inline lib/api/src/lib/stories.ts]: try to get rid of this and split the ref in the router

File: lib/api/src/modules/refs.ts [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+      store.setState({
+        refs: {
+          ...(store.getState().refs || {}),
+          [id]: result,
+        },
+      });
+    },
+  };
+
+  const refs = Object.entries(api.getRefs());
```
Comment (tmeasday) [inline lib/api/src/modules/refs.ts]:
```suggestion
  const refs = Object.entries(provider.getConfig().refs);
```

This is the point where we move the static refs into state.

File: lib/api/src/modules/refs.ts [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+  const initialState: SubState['refs'] = refs.reduce(
+    (acc, [key, data]) => ({
+      ...acc,
+      [key]: {
+        title: data.id,
+        ...data,
+      },
+    }),
+    {} as SubState['refs']
+  );
```
Comment (tmeasday) [inline lib/api/src/modules/refs.ts]: Do you think we should use [`lodash.mapValues`](https://lodash.com/docs/4.17.15#mapValues) for this kind of thing? I feel like it would be more readable and also probably more efficient. Or we could make our own version.

File: lib/channel-postmessage/src/index.ts [hunk 11]
```diff
@@ -55,27 +55,43 @@ export class PostmsgTransport {
...
+      } catch (e) {
+        console.error('sending over postmessage fail');
+      }
+    });
+
+    if (!frames.length || this.buffer.length) {
+      return new Promise((resolve, reject) => {
+        this.buffer.push({ event, resolve, reject });
+      });
+    }
```
Comment (tmeasday) [inline lib/channel-postmessage/src/index.ts]: Does this mean we buffer messages until the first frame appears, and then send them all to just that first frame?

We probably need to be careful that rather than the *first* frame, the **local** frame is the one we (a) wait for and (b) send the buffered messages to.

File: lib/channel-postmessage/src/index.ts [hunk 12]
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
...
-      }
-      return iframe.contentWindow;
+      const nodes: HTMLIFrameElement[] = [
+        ...document.querySelectorAll('#storybook-preview-wrapper iframe'),
+      ];
+
+      const list = nodes
+        .filter(e => {
+          try {
+            return !!e.contentWindow && e.dataset.isStorybook !== undefined && e.id === target;
```
Comment (tmeasday) [inline lib/channel-postmessage/src/index.ts]: Does this code work if `target` is undefined? Can you write a comment explaining what it is doing?

File: lib/channel-postmessage/src/index.ts [hunk 13]
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
...
     }
-    return window.parent;
+    if (window && window.parent) {
+      return [window.parent];
+    }
+
+    return [];
+  }
+
+  private getCurrentFrame(): Window[] {
```
Comment (tmeasday) [inline lib/channel-postmessage/src/index.ts]: Seems odd that this returns an array.

File: lib/channel-postmessage/src/index.ts [hunk 14]
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
...
+    if (this.config.page === 'manager') {
+      const list: HTMLIFrameElement[] = [
+        ...document.querySelectorAll('#storybook-preview-wrapper span + iframe'),
```
Comment (tmeasday) [inline lib/channel-postmessage/src/index.ts]: I don't like this implementation detail from the `Preview` component leaking into the postmessage.

File: lib/channel-postmessage/src/index.ts [hunk 15]
```diff
@@ -89,25 +105,56 @@ export class PostmsgTransport {
...
+      ];
+      return list.map(e => e.contentWindow);
```
Comment (tmeasday) [inline lib/channel-postmessage/src/index.ts]: What if this throws? (CORS)

File: lib/channel-postmessage/src/index.ts [hunk 16]
```diff
@@ -117,6 +164,39 @@ export class PostmsgTransport {
...
+    result = origin + pathname;
+  } catch (e) {
+    // if the contentWindow access was not allowed we match on origin alone
+    [frame, ...remainder] = frames.filter(element => {
+      const src = element.getAttribute('src');
+      const { origin } = new URL(src);
+
+      return origin === event.origin;
+    });
+  }
```
Comment (tmeasday) [inline lib/channel-postmessage/src/index.ts]: It seems a bit weird that we will immediately fallback to the second method if *any* of the iframes don't allow CORS access. What about something more like the following:

```js
const [frame, ...remainder] = frames.filter(element => {
  try {
    return element.contentWindow === event.source;
  } catch (err) {
    // typical case, the iframe has not redirected at all
    return element.getAttribute('src') === event.source.document.location.href
  }
});

if (remainder.length > 0) throw new Error("something's weird");

if (frame) return [frame]

// We couldn't find the frame matching on window or current URL, see if there is a single frame with matching origin
const [frame, ...remainder] = frames.filter(e => new URL(element.getAttribute('src')).origin === event.origin);
if (frame && remainder.length === 0) return frame;

return null;
```

WDYT?

File: lib/core/src/server/manager/manager-webpack.config.js [hunk 17]
```diff
@@ -46,6 +48,17 @@ export default ({
...
+        ? new VirtualModulePlugin({
+            [path.resolve(path.join(configDir, `generated-refs.js`))]: `
+              import { addons } from '@storybook/addons';
+
+              addons.setConfig({
+                refs: ${JSON.stringify(refs)},
+              });
+            `,
+          })
+        : null,
```
Comment (tmeasday) [inline lib/core/src/server/manager/manager-webpack.config.js]: We might want to eventually generalize passing static data from `main.js` to the manager (and the preview for that matter), but this is 👍 for now.

File: lib/ui/src/components/preview/FramesRenderer.tsx [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+    setFrames(newFrames);
+  }, [storyId, story, refs]);
+
+  return (
+    <Fragment>
+      <Global styles={styles} />
+      {Object.entries(frames).map(([id, src]) => (
+        <Fragment key={id}>
+          {id === active ? <span key={`${id}-indicator`} /> : null}
```
Comment (tmeasday) [inline lib/ui/src/components/preview/FramesRenderer.tsx]: As discussed I reckon we can probably just use a `useEffect` in `Iframe` to achieve this.

**Review by tmeasday — CHANGES_REQUESTED** [review]
A few comments; I'd like to take another look after splitting manager-side storyIds in to `{ ref, storyId }`.

I didn't really review the `lib/ui` changes as they are pretty huge but it looks great!

File: lib/api/src/modules/refs.ts [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+export interface SubAPI {
+  findRef: (source: string) => InceptionRef;
+  setRef: (id: string, data: SetRefData, ready?: boolean) => void;
+  getRefs: () => Refs;
+  checkRef: (ref: SetRefData) => Promise<void>;
+  changeRefVersion: (id: string, url: string) => void;
+}
+
+export type Mapper = (ref: InceptionRef, story: StoryInput) => StoryInput;
+export interface InceptionRef {
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/api/src/modules/refs.ts]: Yeah I'll rename it to ComposedRef, I din't want to name this Ref, because that might conflict with React.Ref in places

File: lib/api/src/index.tsx [hunk 20]
```diff
@@ -18,37 +18,26 @@ import {
...
+import * as provider from './modules/provider';
+import * as addons from './modules/addons';
+import * as channel from './modules/channel';
+import * as notifications from './modules/notifications';
+import * as stories from './modules/stories';
+import * as refs from './modules/refs';
+import * as layout from './modules/layout';
+import * as shortcuts from './modules/shortcuts';
+import * as url from './modules/url';
+import * as version from './modules/versions';
```
Comment (ndelangen) [inline lib/api/src/index.tsx]: I think this nicely cleans up the file

File: lib/api/src/modules/refs.ts [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+        api.setRef(id, { id, url, startInjected: true });
+      }
+    },
+
+    getRefs: () => {
+      const { refs: fromConfig = {} } = provider.getConfig();
+      const { refs: fromState = {} } = store.getState();
+
+      return { ...fromConfig, ...fromState };
+    },
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/api/src/modules/refs.ts]: ✅

File: lib/api/src/modules/refs.ts (hunk 10 above)
Comment (ndelangen) [reply lib/api/src/modules/refs.ts]: turns out this is already guarded against in lib/core

File: lib/channel-postmessage/src/index.ts (hunk 14 above)
Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: agreed, will improve

File: lib/channel-postmessage/src/index.ts (hunk 15 above)
Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: it can't, if accessing `.contentWindow` would throw, it has already thrown in the filter.

File: lib/channel-postmessage/src/index.ts (hunk 16 above)
Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: I tweaked to WDYT of it now?

File: lib/core/src/server/manager/manager-webpack.config.js (hunk 17 above)
Comment (ndelangen) [reply lib/core/src/server/manager/manager-webpack.config.js]: agreed

File: lib/channel-postmessage/src/index.ts (hunk 16 above)
Comment (tmeasday) [reply lib/channel-postmessage/src/index.ts]: It’s better but I still think we could check the `document.location.href` inside the iframe. Would that not work?

Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: that's not valid according to the types of the EventSource, which is why I didn't do it..

File: lib/api/src/modules/refs.ts [hunk 22]
```diff
@@ -0,0 +1,205 @@
...
+            ),
+            ref
+          )
+        : undefined;
+
+      const result = { ...ref, stories: after, ...rest, ready };
+
+      store.setState({
+        refs: {
+          ...(store.getState().refs || {}),
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply lib/api/src/modules/refs.ts]: ✅

File: lib/channel-postmessage/src/index.ts (hunk 16 above)
Comment (tmeasday) [reply lib/channel-postmessage/src/index.ts]: I don't really get that? [MDN says](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) that `event.source` is:

> A reference to the window object that sent the message; you can use this to establish two-way communication between two windows with different origins.

So can't we use that to get a reference to the location? Haven't tried it obviously.

File: lib/cli/src/extract.js [hunk 23]
```diff
@@ -0,0 +1,88 @@
+import puppeteer from 'puppeteer';
```
Comment (tmeasday) [inline lib/cli/src/extract.js]: Can we pull this into a separate PR?

File: lib/router/src/utils.ts [hunk 24]
```diff
@@ -15,22 +15,25 @@ export const { toId, parseKind, sanitize, storyNameFromExport } = {
...
+export interface StoryData {
   viewMode?: string;
   storyId?: string;
+  refId?: string;
 }
 
-const splitPathRegex = /\/([^/]+)\/([^/]+)?/;
+const splitPathRegex = /\/([^/]+)\/(?:(.*)_)?([^/]+)?/;
 
 export const parsePath: (path?: string) => StoryData = memoize(1000)(
```
Comment (tmeasday) [inline lib/router/src/utils.ts]: Can we have a test for this?

Comment (ndelangen) [reply lib/router/src/utils.ts]: yes

File: lib/cli/src/extract.js (hunk 23 above)
Comment (ndelangen) [reply lib/cli/src/extract.js]: sure

File: lib/channel-postmessage/src/index.ts (hunk 11 above)
Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: I changed it so it's guaranteed that the local/main frame will get the flush.

File: lib/channel-postmessage/src/index.ts (hunk 12 above)
Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: if target is undefined, it will get the currentFrame:

> `return list.length ? list : this.getCurrentFrame();`

File: lib/channel-postmessage/src/index.ts (hunk 13 above)
Comment (ndelangen) [reply lib/channel-postmessage/src/index.ts]: Seems quite likely to me, we'll have multiple active frames at a time in the future, I'll make this function name plural.

File: lib/cli/src/extract.js (hunk 23 above)
Comment (ndelangen) [reply lib/cli/src/extract.js]: https://github.com/storybookjs/storybook/pull/10146

**Comment by mrmckeb** [comment]
This looks really cool @ndelangen - it's a huge PR, but I get that a lot of this work wouldn't make sense in isolation.

I can already see a lot of use-cases for this, from versioning through to framework mixing, and of course importing Storybooks from other sources (such as consumed libraries).

@tmeasday, perhaps it's best if we get this in now, then create seperate PR(s) for authentication and UI updates? I think the sooner we get people playing with this, the better. I also feel like there will be quite a few tweaks and changes based on community feedback.

**Comment by tmeasday** [comment]
> @tmeasday, perhaps it's best if we get this in now, then create seperate PR(s) for authentication and UI updates? I think the sooner we get people playing with this, the better. I also feel like there will be quite a few tweaks and changes based on community feedback.

Yeah, I'm down with that. There were a few outstanding things on the PR, I'm not sure if @ndelangen has addressed them yet? (It's such a beast it's hard to get a quick overview).

In any case it's minor stuff now.

**Comment by mrmckeb** [comment]
It looked like most had been addressed when I was reviewing yesterday. Norbert walked me through it which helped. @ndelangen was there anything else?
